src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -534,44 +526,88 @@ private static async Task WriteDataAsync(Stream archiveStream, Stream dataStream
             cancellationToken.ThrowIfCancellationRequested();
 
             await dataStream.CopyToAsync(archiveStream, cancellationToken).ConfigureAwait(false); // The data gets copied from the current position
+
             int paddingAfterData = TarHelpers.CalculatePadding(actualLength);
-            await archiveStream.WriteAsync(new byte[paddingAfterData], cancellationToken).ConfigureAwait(false);
+            if (paddingAfterData != 0)
+            {
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(paddingAfterData);
```

Comment:
The assert that was added in the sync version:
```suggestion
                Debug.Assert(paddingAfterData <= TarHelpers.RecordSize);

                byte[] buffer = ArrayPool<byte>.Shared.Rent(paddingAfterData);
```